popup/search-results.js

```diff
@@ -0,0 +1,547 @@
+/* global handleEvent tryJSONparse getStylesSafe BG */
+'use strict';
+
+/**
+ * Library for interacting with userstyles.org
+ * @returns {Object} Exposed methods representing the search results on userstyles.org
+ */
+function SearchUserstyles() {
+  let totalPages;
+  let currentPage = 1;
+  let exhausted = false;
+
+  return {getCurrentPage, getTotalPages, getCategory, isExhausted, search, fetchStyleJson, fetchStyle};
+
+  /**
+   * @returns {Number} The *Next* page to fetch for styles. Auto-incremented after each search.
+   */
+  function getCurrentPage() {
+    return currentPage;
+  }
+
+  /**
+   * @returns {Number} The total number of search result pages.
+   */
+  function getTotalPages() {
+    return totalPages;
+  }
+
+  /**
+   * @returns {Boolean} If there are no more results to fetch from userstyles.org
+   */
+  function isExhausted() {
+    return exhausted;
+  }
+
+  /**
+   * Resolves the Userstyles.org "category" for a given URL.
+   * @param {String} url The URL to a webpage.
+   * @returns {Promise<String>} The category for a URL, or the hostname if category is not found.
+   */
+  function getCategory(url) {
+    const hostname = new URL(url).hostname;
+    return new Promise(resolve => {
+      const request = new XMLHttpRequest();
+      const browseURL = 'https://userstyles.org/styles/browse/all/' + encodeURIComponent(url);
+      request.open('HEAD', browseURL, true);
```

Comment:
you can use `download()` - it accepts custom request parameters now; the base URL should be a named constant at the top of `searchUserstyles` closure